### PR TITLE
Add more tests for homie-controller

### PR DESCRIPTION
**Pull request by qwandor-google**
Better late than never.

### CODE REVIEW

File: homie-controller/src/lib.rs [hunk 1]
```diff
@@ -617,3 +616,206 @@ impl From<ParseFloatError> for HandleError {
...
+            Subscribe::new("base_topic/device_id/+", QoS::AtLeastOnce),
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/$fw/+", QoS::AtLeastOnce),
+        );
+        expect_subscribe(
+            requests_rx.recv().await?,
+            Subscribe::new("base_topic/device_id/$stats/+", QoS::AtLeastOnce),
+        );
```
Comment (alsuren) [inline homie-controller/src/lib.rs:690]: (crazy idea): Could we have something like this that doesn't care about ordering?
```suggestion
        expect_subscriptions(
            requests_rx,
            vec!["base_topic/device_id/+", "base_topic/device_id/$fw/+", "base_topic/device_id/$stats/+"],
        );
```

File: homie-controller/src/lib.rs [hunk 2]
```diff
@@ -620,3 +619,206 @@ impl From<ParseFloatError> for HandleError {
...
+            base_topic: "base_topic".to_owned(),
+            mqtt_client,
+            devices: Mutex::new(Arc::new(HashMap::new())),
+        };
+        (controller, requests_rx)
+    }
+
+    fn expect_subscribe(request: Request, expected: Subscribe) {
+        if let Request::Subscribe(actual) = request {
+            assert_eq!(actual, expected);
```
[the remark this answers is not in the document]
Comment (alsuren) [reply homie-controller/src/lib.rs:640]: Also, it looks like we always subscribe with QoS::AtLeastOnce in tests, so this helper could just hard-code it.

File: homie-controller/src/types.rs [hunk 3]
```diff
@@ -824,4 +840,63 @@ mod tests {
...
+        assert_eq!(node.has_required_attributes(), false);
+
+        node.name = Some("Node name".to_owned());
+        assert_eq!(node.has_required_attributes(), false);
+
+        node.node_type = Some("Node type".to_owned());
+        assert_eq!(node.has_required_attributes(), false);
+
+        node.add_property(property_with_required_attributes());
+        assert_eq!(node.has_required_attributes(), true);
```
Comment (alsuren) [inline homie-controller/src/types.rs:876]: so node needs at least one property? I suppose that makes sense.

File: homie-controller/src/types.rs [hunk 4]
```diff
@@ -824,4 +840,63 @@ mod tests {
...
+        assert_eq!(device.has_required_attributes(), false);
+
+        device.name = Some("Device name".to_owned());
+        assert_eq!(device.has_required_attributes(), false);
+
+        device.state = State::Init;
+        assert_eq!(device.has_required_attributes(), true);
+
+        device.add_node(node_with_required_attributes());
+        assert_eq!(device.has_required_attributes(), true);
```
Comment (alsuren) [inline homie-controller/src/types.rs:900]: if you add a node without the required attributes, does it go to false?

File: homie-controller/src/lib.rs [hunk 5]
```diff
@@ -617,3 +616,206 @@ impl From<ParseFloatError> for HandleError {
...
+                "Node type"
+            )
+            .await?,
+            Some(Event::NodeUpdated {
+                device_id: "device_id".to_owned(),
+                node_id: "node_id".to_owned(),
+                has_required_attributes: false
+            })
+        );
+        let mut expected_node = Node::new("node_id");
```
Comment (alsuren) [inline homie-controller/src/lib.rs:751]: I still don't like these variables that evolve throughout the test. Let me try something.

File: homie-controller/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (qwandor) [reply homie-controller/src/lib.rs:640]: 'Request' unfortunately does not currently derive `PartialEq`, which is why I added this helper function. I sent https://github.com/bytebeamio/rumqtt/pull/171 to fix that; once that gets released we can clean this up.

File: homie-controller/src/lib.rs (hunk 1 above)
Comment (qwandor) [reply homie-controller/src/lib.rs:690]: Good idea, I'll do that.

File: homie-controller/src/types.rs (hunk 3 above)
Comment (qwandor) [reply homie-controller/src/types.rs:876]: That's the current logic, yes. Whether that's how it should be I'm not entirely sure, and I'm inconsistent with devices.

File: homie-controller/src/types.rs (hunk 4 above)
Comment (qwandor) [reply homie-controller/src/types.rs:900]: Added.

**Pull request merged**